### DISCUSSION TITLE
Linux gfortran-13 code missing library

### DISCUSSION THREAD

**Discussion by acombescure** [Homebrew/discussions#4473] [Linux]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: 01e0c20d01c6aab28d844280d130f0e745e2ddd9
Last commit: 2 days ago
Core tap JSON: 01 May 20:52 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.QfXfe7D7TZ/org.xquartz:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

i have compiled and linked with gfortran-13 uploaded with the brew command brew install gfortran
a fortran program of about 1000 subroutines that works on all machines since 40 Years. 
It was nicely working on my previous mac (OS 10)
the compilation and link stage have no problem

but when i try to launch the program i get a library problem. described in the following lines...

dyld[34032]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
dyld[34032]: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: <281461F9-AB34-36A4-9311-FF47FC96240A> /Users/alaincombescure/Documents/Stanlax/dinca/coq/incaessai
  Reason: tried: '/usr/lib/libSystem.B.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/libSystem.B.dylib'

what can i do to overcome this apparently missing library  problem ?(libSystem.B.dylib not found) 

thanks for the help..

**Comment by SMillerDev** [comment]
Afaik GCC on ARM macs is still experimental and officially unsupported so you'd have to report this to https://github.com/iains/gcc-13-branch who created the patch we're using.

**Comment by fxcoudert** [comment]
Can you compile and run this short test program?

```
meau /tmp $ cat a.f90 
  real :: x = 2.
  print *, sqrt(x)
end
meau /tmp $ gfortran a.f90 && ./a.out
   1.41421354    
meau /tmp $ otool -L ./a.out         
./a.out:
	/opt/homebrew/opt/gcc/lib/gcc/current/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/opt/homebrew/opt/gcc/lib/gcc/current/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

**Comment by acombescure** [comment]
Thanks … I get the following parse error….
Sorry but I do not know what to do…

***@***.*** coq % meau /tmp $ cat a.f90                
  real :: x = 2.
  print *, sqrt(x)
end
meau /tmp $ gfortran a.f90 && ./a.out
   1.41421354
meau /tmp $ otool -L ./a.out
./a.out:
        /opt/homebrew/opt/gcc/lib/gcc/current/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
        /opt/homebrew/opt/gcc/lib/gcc/current/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
zsh: parse error near `end’

Regards

**Reply by fxcoudert** [reply]
Create a file named `a.f90` with the following content:

```
  real :: x = 2.
  print *, sqrt(x)
end
```

Then run `gfortran a.f90 && ./a.out`
Then run `otool -L ./a.out`

**Comment by acombescure** [comment]
Thanks,

It works 
Here is the result

The program seems to find the library

***@***.*** coq % gfortran a.f90 && ./a.out
   1.41421354    
***@***.*** coq % otool -L ./a.out
./a.out:
	/opt/homebrew/opt/gcc/lib/gcc/current/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/opt/homebrew/opt/gcc/lib/gcc/current/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
It also works with gfortran-13

Regards

Alain

**Reply by fxcoudert** [reply]
So it works on a simple code, meaning the compiler is functional. Why it fails on your real example, I cannot tell, you would have to reduce it to a simpler example that we can reproduce. The real error is this line:

```
dyld[34032]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
```

`dyld` should be able to load its cache, and the cache should not be `null`. Why that occurs, I have no idea.

**Comment by acombescure** [comment]
thanks
Here is the linking phase shell
#! /bin/bash
echo ">>>>>  edition de liens de INCA <<<<"
####################   gfortran
time gfortran-13 -O -o./incaessai  /Users/alaincombescure/Documents/Stanlax/castem/inca/main.o  /Users/alaincombescure/Documents/Stanlax/castem/inca/inca.a

chmod +x ./incaessai
exit

Do you see something wrong 

Regards

Alain

**Reply by fxcoudert** [reply]
No, I do not seen anything wrong with that. In 90% of errors I can get a feeling of what went wrong, or broadly where the issue is. But in your case it's a new one, I've never seen it. It's either something weird, or something whose cause is deep.

Given it's a dyld cache issue, among the “weird” category: are you running macOS in a weird state? (recovery mode, or with SIP protection disabled)

**Comment by acombescure** [comment]
Thanks
No I am not on a recovery mode…nor did disabled SIP protection .. (at least I have not done this on purpose)
Regards

**Comment by acombescure** [comment]
I have found the reason for dylib  cache .. Rather simple 
I had chosen to reserve a place for memory of 500 000 000 R*8
When I diminishes this value to 250 000 000 no more library problem
This is annoying.. on Mac OS X 500 000 000 was OK

Regards

Alain

**Comment by fxcoudert** [comment]
You can allocate that amount dynamically, it works:

```
meau /tmp $ cat a.f90 
  real(kind=8), allocatable :: x(:)
  allocate(x(500000000))
  x(1) = 2
  print *, sqrt(x(1))
  print *, sizeof(x) / 1024 / 1024, 'MB'
end
meau /tmp $ gfortran a.f90 && ./a.out 
   1.4142135623730951     
                 3814 MB
```

**Comment by acombescure** [comment]
Thanks a lot
I will do this
Regards

Alain
